crates/ide_db/src/search.rs

```diff
@@ -52,23 +61,53 @@ impl IntoIterator for UsageSearchResult {
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum NameLike {
+    NameRef(ast::NameRef),
+    Name(ast::Name),
+    Lifetime(ast::Lifetime),
+}
+
+mod __ {
+    use super::{
+        ast::{Lifetime, Name, NameRef},
+        NameLike,
+    };
+    stdx::impl_from!(NameRef, Name, Lifetime for NameLike);
+}
+
 #[derive(Debug, Clone)]
 pub struct FileReference {
     pub range: TextRange,
-    pub kind: ReferenceKind,
+    pub name: NameKind,
     pub access: Option<ReferenceAccess>,
 }
 
-#[derive(Debug, Clone, PartialEq)]
-pub enum ReferenceKind {
-    FieldShorthandForField,
-    FieldShorthandForLocal,
-    StructLiteral,
-    RecordFieldExprOrPat,
-    SelfParam,
-    EnumLiteral,
-    Lifetime,
-    Other,
+impl FileReference {
+    pub fn name_from_syntax(&self, root: &SyntaxNode) -> Option<NameLike> {
+        let node = node_or_parent(root.covering_element(self.range));
+        match self.name {
+            NameKind::Name => ast::Name::cast(node).map(Into::into),
+            NameKind::NameRef => ast::NameRef::cast(node).map(Into::into),
+            NameKind::Lifetime => ast::Lifetime::cast(node).map(Into::into),
+        }
+    }
+
+    pub fn as_name_ref(&self, root: &SyntaxNode) -> Option<ast::NameRef> {
+        match self.name {
+            NameKind::NameRef => {
+                ast::NameRef::cast(node_or_parent(root.covering_element(self.range)))
+            }
+            _ => None,
+        }
+    }
+}
```

Comment:
@matklad Okay I changed `FileReference` so that it only contains the `TextRange` of the name now and what kind it is, as in whether it is a `NameRef`, `Name` or `Lifetime`. Instead it has these methods now, with this there are no `AST` nodes being kept in the fields anymore so this should be alright if I am not mistaken? Though `NameLike` should probably be defined in `syntax` now or so as it is still required publicly.

Reply:
I just realized having the ast node in FileReference is fine now since `ide_db` can expose that just fine right? the only problem was that it was re-exposed in `ide` through the `ReferenceSearchResult`.